src/main/scala/mesosphere/marathon/core/volume/providers/DVDIProvider.scala

```diff
@@ -0,0 +1,243 @@
+  // for now this matches the validation for resident tasks, but probably won't be as
+  // restrictive in the future.
+  private val validUpgradeStrategy = validator[UpgradeStrategy] { strategy =>
+    strategy.minimumHealthCapacity should be <= 0.5
+    strategy.maximumOverCapacity should equalTo(0.0)
+  }
+
+  /** @return a count of volume references-by-name within an app spec */
+  private def volumeNameCounts(app: AppDefinition): Map[String, Int] =
+    volumesForApp(app).flatMap{ v => nameOf(v.external) }.groupBy(identity).mapValues(_.size)
+
+  protected[providers] def modes(ct: Container): Set[Mode] =
+    collect(ct).map(_.mode).toSet
+
+  private val validMesosVolume = validator[ExternalVolume] { v =>
+    v.mode is equalTo(Mode.RW)
+  }
+
+  private val validMesosContainer = validator[Container] { ct =>
+    ct.volumes.each is ifDVDIVolume(validMesosVolume)
+  }
+
+  private val haveOnlyOneDriver = new Validator[Container] {
+    override def apply(ct: Container): Result = {
+      val n = collect(ct).flatMap(_.external.options.get(driverOption)).toSet.size
+      if (n <= 1) Success
+      else Failure(Set(RuleViolation(n, "one external volume driver per app", None)))
+    }
+  }
+
+  private val haveOnlyDriverOption = new Validator[ExternalVolume] {
+    override def apply(v: ExternalVolume): Result =
+      if (v.external.options.filterKeys(_ != driverOption).isEmpty) Success
+      else Failure(Set(RuleViolation(v.external.options, "only contains driver", Some("external.options"))))
+  }
+
+  private val haveNoSize = new Validator[ExternalVolume] {
+    override def apply(v: ExternalVolume): Result =
+      if (v.external.size.isEmpty) Success
+      else Failure(Set(RuleViolation(v.external.size, "must be undefined", Some("external/size"))))
+  }
+
+  private def ifDVDIVolume(vtor: Validator[ExternalVolume]) = new Validator[Volume] {
+    override def apply(v: Volume): Result = v match {
+      case ev: ExternalVolume if accepts(ev) => vtor(ev)
+      case _                                 => Success
+    }
+  }
+
+  private val validDockerContainer = validator[Container] { ct =>
+    ct should haveOnlyOneDriver
+    ct.volumes.each should ifDVDIVolume(haveOnlyDriverOption)
+    ct.volumes.each should ifDVDIVolume(haveNoSize)
+  }
+
+  private val validContainer = new Validator[Container] {
+    override def apply(ct: Container): Result = ct.`type` match {
+      case ContainerInfo.Type.MESOS  => validMesosContainer(ct)
+      case ContainerInfo.Type.DOCKER => validDockerContainer(ct)
+    }
+  }
+
+  val appValidation = validator[AppDefinition] { app =>
+    app should haveUniqueExternalVolumeNames
+    app should haveOnlyOneReplica
+    app.container is ifDefined(validContainer)
+    app.upgradeStrategy is validUpgradeStrategy
+  }
+
+  // group-level validation for DVDI volumes: the same volume name may only be referenced by a single
+  // task instance across the entire cluster.
+  val groupValidation = new Validator[Group] {
```

Comment:
I would probably rename this (and super and dependent names) to "rootGroup" validation and only call it on a rootGroup.